stl/inc/semaphore

```diff
@@ -0,0 +1,85 @@
+// semaphore standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+
+#ifndef _SEMAPHORE_
+#define _SEMAPHORE_
+#include <yvals_core.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#if !_HAS_CXX20
+#pragma message("The contents of <semaphore> are available only with C++20 or later.")
+#else // ^^^ !_HAS_CXX20 / _HAS_CXX20 vvv
+#include <atomic>
+#include <cassert>
+#include <chrono>
+#include <cstddef>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+template <ptrdiff_t _LeastMaxValue = 256>
+class counting_semaphore {
+public:
+    static constexpr ptrdiff_t max() noexcept {
+        // _LeastMaxValue is used here so that things such as binary_semaphore are
+        return _LeastMaxValue;
+    }
+    constexpr explicit counting_semaphore(ptrdiff_t _Desired) {
+        _STL_ASSERT(_Desired >= 0 && _Desired <= max(), "Desired value for the semaphore falls outside of semaphore range");
+        atomic_init(&_Counter, _Desired);
+    }
+    ~counting_semaphore()                         = default;
+    counting_semaphore(const counting_semaphore&) = delete;
+    counting_semaphore& operator=(const counting_semaphore&) = delete;
+
+    void release(ptrdiff_t _Update = 1) {
+        _STL_ASSERT(_Update >= 0 && _Update <= (max() - _Counter.load()), "The semaphore has been released more times than acquired");
+        _Counter.fetch_add(_Update, memory_order_release);
+    }
+    void acquire() {
+        while (!try_acquire()) {
+        }
+    }
+    bool try_acquire() noexcept {
+        ptrdiff_t _Old     = _Counter.load(memory_order_acquire);
+        ptrdiff_t _Desired = _Old - 1;
+        if (_Old > 0 && _Counter.compare_exchange_strong(_Old, _Desired, memory_order_acquire)) {

```

Comment:
You didn't recalculate `_Desired` if the CAS fails. The usual CAS loop for a subtraction would look like:

```
ptrdiff_t _Old = 0;
ptrdiff_t _Desired;
do {
    _Desired = _Old - 1;
while (_The_atomic.compare_exchange_weak(_Old, _Desired));
```

though like I said without waiting / backoff this is insufficient. (I haven't thought about getting the semaphore parts connected to this but I'm not working on it right now :) )

Reply:
The standard requires that if the CAS fails that we just move on for try_acquire so I didn't put a loop there in the first place since if you want blocking semantics you're supposed to go to try_acquire_for, try_acquire_until, and acquire.

Reply:
This isn't a "block until the semaphore is available" loop, this is a "update the value only if another CPU didn't update it in the meantime" loop.

Reply:
But wouldn't this loop still prevent try_acquire() from returning false in any condition or am I missing something while going through the atomic_compare_exchange pages?

Reply:
The loop I wrote here just decrements a number to demonstrate how to use CAS, not how to implement try_acquire; the only time the loop repeats is when some other thread changed the value at the exact same time this thread did; thus it has guaranteed forward progress (that is, the only way this thread doesn't make forward progress is if some other thread made forward progress).

Or is the confusion that you missed that CAS loads the actual value into `_Old`?

Reply:
I was confused because I thought that it made doing a proper return if false or true for the try_acquire impossible, I'm still confused how the loop allows for this at all, because we need to somehow save the result from the return of the compare_exchange_strong.